File: backend/src/routes/reports.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const pool = require('../db');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();

// GET /reports
router.get('/', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    let result;

    if (user.role === 'patient') {
      result = await pool.query(
        `SELECT * FROM medical_reports WHERE patient_id = $1 ORDER BY upload_date DESC`,
        [user.id]
      );
    } else if (user.role === 'doctor') {
      result = await pool.query(
        `SELECT * FROM medical_reports WHERE doctor_id = $1 ORDER BY upload_date DESC`,
        [user.id]
      );
    } else {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    const reports = result.rows.map(mapReport);
    return res.json({
      reports,
      totalCount: reports.length,
      analyzedCount: reports.filter(r => r.aiAnalysis).length,
      pendingCount: reports.filter(r => !r.aiAnalysis).length,
    });
  } catch (err) {
    console.error('Fetch reports error:', err);
    return res.status(500).json({ error: 'Failed to fetch reports' });
  }
});

// GET /reports/:id
router.get('/:id', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    const { id } = req.params;

    const result = await pool.query(`SELECT * FROM medical_reports WHERE id = $1`, [id]);
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Report not found' });
    }

    const r = result.rows[0];
    if (r.patient_id !== user.id && r.doctor_id !== user.id) {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    return res.json(mapReport(r));
  } catch (err) {
    console.error('Fetch report error:', err);
    return res.status(500).json({ error: 'Failed to fetch report' });
  }
});

// POST /reports - upload a report (patient only)
router.post('/', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    if (user.role !== 'patient') {
      return res.status(403).json({ error: 'Only patients can upload reports' });
    }

    const { doctorId, reportType, labName, cost, fileUrl, aiAnalysis } = req.body;

    // Get doctor name if provided
    let doctorName = null;
    if (doctorId) {
      const docRes = await pool.query(`SELECT full_name FROM users WHERE id = $1`, [doctorId]);
      if (docRes.rows.length > 0) doctorName = docRes.rows[0].full_name;
    }

    const id = uuidv4();
    const result = await pool.query(
      `INSERT INTO medical_reports (
        id, patient_id, doctor_id, patient_name, doctor_name,
        report_type, lab_name, cost, file_url, ai_analysis
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)
      RETURNING *`,
      [
        id, user.id, doctorId || null,
        user.full_name, doctorName,
        reportType || null, labName || null, cost || null,
        fileUrl || null,
        aiAnalysis ? JSON.stringify(aiAnalysis) : null,
      ]
    );

    // Log activity for doctor if provided
    if (doctorId) {
      await pool.query(
        `INSERT INTO doctor_activity (id, doctor_id, patient_id, patient_name, activity_type, report_type)
         VALUES ($1,$2,$3,$4,'report_upload',$5)`,
        [uuidv4(), doctorId, user.id, user.full_name, reportType]
      );
    }

    return res.status(201).json({
      success: true,
      report: mapReport(result.rows[0]),
    });
  } catch (err) {
    console.error('Upload report error:', err);
    return res.status(500).json({ error: 'Failed to upload report' });
  }
});

// DELETE /reports/:id - delete a report (patient only)
router.delete('/:id', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    const { id } = req.params;

    const check = await pool.query(
      `SELECT id FROM medical_reports WHERE id = $1 AND patient_id = $2`,
      [id, user.id]
    );
    if (check.rows.length === 0) {
      return res.status(404).json({ error: 'Report not found or unauthorized' });
    }

    await pool.query(`DELETE FROM medical_reports WHERE id = $1`, [id]);
    return res.json({ success: true });
  } catch (err) {
    console.error('Delete report error:', err);
    return res.status(500).json({ error: 'Failed to delete report' });
  }
});

// PUT /reports/:id - update AI analysis
router.put('/:id', requireAuth, async (req, res) => {
  try {
    const user = req.user;
    const { id } = req.params;
    const { aiAnalysis } = req.body;

    const result = await pool.query(
      `UPDATE medical_reports SET ai_analysis = $1 WHERE id = $2 AND patient_id = $3 RETURNING *`,
      [JSON.stringify(aiAnalysis), id, user.id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Report not found' });
    }

    return res.json({ success: true, report: mapReport(result.rows[0]) });
  } catch (err) {
    console.error('Update report error:', err);
    return res.status(500).json({ error: 'Failed to update report' });
  }
});

function mapReport(row) {
  return {
    id: row.id,
    patientId: row.patient_id,
    doctorId: row.doctor_id,
    patientName: row.patient_name,
    doctorName: row.doctor_name,
    reportType: row.report_type,
    labName: row.lab_name,
    cost: row.cost,
    fileUrl: row.file_url,
    aiAnalysis: row.ai_analysis,
    uploadDate: row.upload_date,
    createdAt: row.created_at,
  };
}

module.exports = router;

File: backend/routers/reports.py
"""
routers/reports.py — Medical report CRUD routes for Clynicx Python backend.
Mirrors the behaviour of the original Node.js src/routes/reports.js.

Endpoints:
    GET    /reports
    GET    /reports/{report_id}
    POST   /reports
    DELETE /reports/{report_id}
    PUT    /reports/{report_id}
"""
import uuid
import json
from typing import Optional, Any
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from db import query
from middleware.auth import require_auth

router = APIRouter(prefix="/reports", tags=["reports"])


def _map_report(row: dict) -> dict:
    ai_analysis = row.get("ai_analysis")
    if isinstance(ai_analysis, str):
        try:
            ai_analysis = json.loads(ai_analysis)
        except Exception:
            pass

    return {
        "id": str(row["id"]),
        "patientId": str(row["patient_id"]),
        "doctorId": str(row["doctor_id"]) if row.get("doctor_id") else None,
        "patientName": row.get("patient_name"),
        "doctorName": row.get("doctor_name"),
        "reportType": row.get("report_type"),
        "labName": row.get("lab_name"),
        "cost": row.get("cost"),
        "fileUrl": row.get("file_url"),
        "aiAnalysis": ai_analysis,
        "uploadDate": str(row["upload_date"]) if row.get("upload_date") else None,
        "createdAt": str(row["created_at"]) if row.get("created_at") else None,
    }


# ── Request Bodies ────────────────────────────────────────────────────────────

class UploadReportBody(BaseModel):
    doctorId: Optional[str] = None
    reportType: Optional[str] = None
    labName: Optional[str] = None
    cost: Optional[str] = None
    fileUrl: Optional[str] = None
    aiAnalysis: Optional[Any] = None


class UpdateReportBody(BaseModel):
    aiAnalysis: Any


# ── GET /reports ──────────────────────────────────────────────────────────────

@router.get("")
def list_reports(user: dict = Depends(require_auth)):
    if user["role"] == "patient":
        rows = query(
            "SELECT * FROM medical_reports WHERE patient_id = %s ORDER BY upload_date DESC",
            (user["id"],),
        )
    elif user["role"] == "doctor":
        rows = query(
            "SELECT * FROM medical_reports WHERE doctor_id = %s ORDER BY upload_date DESC",
            (user["id"],),
        )
    else:
        raise HTTPException(status_code=403, detail="Unauthorized")

    reports = [_map_report(r) for r in rows]
    return {
        "reports": reports,
        "totalCount": len(reports),
        "analyzedCount": sum(1 for r in reports if r["aiAnalysis"]),
        "pendingCount": sum(1 for r in reports if not r["aiAnalysis"]),
    }


# ── GET /reports/{report_id} ──────────────────────────────────────────────────

@router.get("/{report_id}")
def get_report(report_id: str, user: dict = Depends(require_auth)):
    rows = query("SELECT * FROM medical_reports WHERE id = %s", (report_id,))
    if not rows:
        raise HTTPException(status_code=404, detail="Report not found")

    r = rows[0]
    if str(r["patient_id"]) != str(user["id"]) and (
        r.get("doctor_id") is None or str(r["doctor_id"]) != str(user["id"])
    ):
        raise HTTPException(status_code=403, detail="Unauthorized")

    return _map_report(r)


# ── POST /reports ─────────────────────────────────────────────────────────────

@router.post("", status_code=201)
def upload_report(body: UploadReportBody, user: dict = Depends(require_auth)):
    if user["role"] != "patient":
        raise HTTPException(status_code=403, detail="Only patients can upload reports")

    # Get doctor name if provided
    doctor_name = None
    if body.doctorId:
        docs = query("SELECT full_name FROM users WHERE id = %s", (body.doctorId,))
        if docs:
            doctor_name = docs[0]["full_name"]

    report_id = str(uuid.uuid4())
    ai_json = json.dumps(body.aiAnalysis) if body.aiAnalysis is not None else None

    rows = query(
        """
        INSERT INTO medical_reports (
            id, patient_id, doctor_id, patient_name, doctor_name,
            report_type, lab_name, cost, file_url, ai_analysis
        ) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
        RETURNING *
        """,
        (
            report_id,
            str(user["id"]),
            body.doctorId or None,
            user["full_name"],
            doctor_name,
            body.reportType,
            body.labName,
            body.cost,
            body.fileUrl,
            ai_json,
        ),
    )

    # Log activity for doctor if provided
    if body.doctorId:
        query(
            """
            INSERT INTO doctor_activity (id, doctor_id, patient_id, patient_name, activity_type, report_type)
            VALUES (%s,%s,%s,%s,'report_upload',%s)
            """,
            (str(uuid.uuid4()), body.doctorId, str(user["id"]), user["full_name"], body.reportType),
        )

    return {"success": True, "report": _map_report(rows[0])}


# ── DELETE /reports/{report_id} ───────────────────────────────────────────────

@router.delete("/{report_id}")
def delete_report(report_id: str, user: dict = Depends(require_auth)):
    check = query(
        "SELECT id FROM medical_reports WHERE id = %s AND patient_id = %s",
        (report_id, str(user["id"])),
    )
    if not check:
        raise HTTPException(status_code=404, detail="Report not found or unauthorized")

    query("DELETE FROM medical_reports WHERE id = %s", (report_id,))
    return {"success": True}


# ── PUT /reports/{report_id} ──────────────────────────────────────────────────

@router.put("/{report_id}")
def update_report(report_id: str, body: UpdateReportBody, user: dict = Depends(require_auth)):
    ai_json = json.dumps(body.aiAnalysis)
    rows = query(
        """
        UPDATE medical_reports
        SET ai_analysis = %s
        WHERE id = %s AND patient_id = %s
        RETURNING *
        """,
        (ai_json, report_id, str(user["id"])),
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Report not found")

    return {"success": True, "report": _map_report(rows[0])}
